File: packages/firebase_common/lib/src/firestore_documents/sample_todo.flutterfire_gen.dart
// coverage:ignore-file
// ignore_for_file: type=lint
// ignore_for_file: unused_element, deprecated_member_use, deprecated_member_use_from_same_package, use_function_type_syntax_for_parameters, unnecessary_const, avoid_init_to_null, invalid_override_different_default_values_named, prefer_expression_function_bodies, annotate_overrides, invalid_annotation_target, unnecessary_question_mark

part of 'sample_todo.dart';

class ReadSampleTodo {
  const ReadSampleTodo({
    required this.sampleTodoId,
    required this.path,
    required this.title,
    required this.description,
    required this.isDone,
    required this.dueDateTime,
    required this.updatedAt,
  });

  final String sampleTodoId;

  final String path;

  final String title;

  final String description;

  final bool isDone;

  final DateTime dueDateTime;

  final DateTime? updatedAt;

  factory ReadSampleTodo._fromJson(Map<String, dynamic> json) {
    return ReadSampleTodo(
      sampleTodoId: json['sampleTodoId'] as String,
      path: json['path'] as String,
      title: json['title'] as String? ?? '',
      description: json['description'] as String? ?? '',
      isDone: json['isDone'] as bool? ?? false,
      dueDateTime: (json['dueDateTime'] as Timestamp).toDate(),
      updatedAt: (json['updatedAt'] as Timestamp?)?.toDate(),
    );
  }

  factory ReadSampleTodo.fromDocumentSnapshot(DocumentSnapshot ds) {
    final data = ds.data()! as Map<String, dynamic>;
    return ReadSampleTodo._fromJson(<String, dynamic>{
      ...data,
      'sampleTodoId': ds.id,
      'path': ds.reference.path,
    });
  }
}

class CreateSampleTodo {
  const CreateSampleTodo({
    required this.title,
    required this.description,
    required this.isDone,
    required this.dueDateTime,
  });

  final String title;
  final String description;
  final bool isDone;
  final DateTime dueDateTime;

  Map<String, dynamic> toJson() {
    return {
      'title': title,
      'description': description,
      'isDone': isDone,
      'dueDateTime': dueDateTime,
      'updatedAt': FieldValue.serverTimestamp(),
    };
  }
}

class UpdateSampleTodo {
  const UpdateSampleTodo({
    this.title,
    this.description,
    this.isDone,
    this.dueDateTime,
  });

  final String? title;
  final String? description;
  final bool? isDone;
  final DateTime? dueDateTime;

  Map<String, dynamic> toJson() {
    return {
      if (title != null) 'title': title,
      if (description != null) 'description': description,
      if (isDone != null) 'isDone': isDone,
      if (dueDateTime != null) 'dueDateTime': dueDateTime,
      'updatedAt': FieldValue.serverTimestamp(),
    };
  }
}

class DeleteSampleTodo {}

/// Provides a reference to the sampleTodos collection for reading.
final readSampleTodoCollectionReference = FirebaseFirestore.instance
    .collection('sampleTodos')
    .withConverter<ReadSampleTodo>(
      fromFirestore: (ds, _) => ReadSampleTodo.fromDocumentSnapshot(ds),
      toFirestore: (_, __) => throw UnimplementedError(),
    );

/// Provides a reference to a sampleTodo document for reading.
DocumentReference<ReadSampleTodo> readSampleTodoDocumentReference({
  required String sampleTodoId,
}) =>
    readSampleTodoCollectionReference.doc(sampleTodoId);

/// Provides a reference to the sampleTodos collection for creating.
final createSampleTodoCollectionReference = FirebaseFirestore.instance
    .collection('sampleTodos')
    .withConverter<CreateSampleTodo>(
      fromFirestore: (_, __) => throw UnimplementedError(),
      toFirestore: (obj, _) => obj.toJson(),
    );

/// Provides a reference to a sampleTodo document for creating.
DocumentReference<CreateSampleTodo> createSampleTodoDocumentReference({
  required String sampleTodoId,
}) =>
    createSampleTodoCollectionReference.doc(sampleTodoId);

/// Provides a reference to the sampleTodos collection for updating.
final updateSampleTodoCollectionReference = FirebaseFirestore.instance
    .collection('sampleTodos')
    .withConverter<UpdateSampleTodo>(
      fromFirestore: (_, __) => throw UnimplementedError(),
      toFirestore: (obj, _) => obj.toJson(),
    );

/// Provides a reference to a sampleTodo document for updating.
DocumentReference<UpdateSampleTodo> updateSampleTodoDocumentReference({
  required String sampleTodoId,
}) =>
    updateSampleTodoCollectionReference.doc(sampleTodoId);

/// Provides a reference to the sampleTodos collection for deleting.
final deleteSampleTodoCollectionReference = FirebaseFirestore.instance
    .collection('sampleTodos')
    .withConverter<DeleteSampleTodo>(
      fromFirestore: (_, __) => throw UnimplementedError(),
      toFirestore: (_, __) => throw UnimplementedError(),
    );

/// Provides a reference to a sampleTodo document for deleting.
DocumentReference<DeleteSampleTodo> deleteSampleTodoDocumentReference({
  required String sampleTodoId,
}) =>
    deleteSampleTodoCollectionReference.doc(sampleTodoId);

/// Manages queries against the sampleTodos collection.
class SampleTodoQuery {
  /// Fetches [ReadSampleTodo] documents.
  Future<List<ReadSampleTodo>> fetchDocuments({
    GetOptions? options,
    Query<ReadSampleTodo>? Function(Query<ReadSampleTodo> query)? queryBuilder,
    int Function(ReadSampleTodo lhs, ReadSampleTodo rhs)? compare,
  }) async {
    Query<ReadSampleTodo> query = readSampleTodoCollectionReference;
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    final qs = await query.get(options);
    final result = qs.docs.map((qds) => qds.data()).toList();
    if (compare != null) {
      result.sort(compare);
    }
    return result;
  }

  /// Subscribes [SampleTodo] documents.
  Stream<List<ReadSampleTodo>> subscribeDocuments({
    Query<ReadSampleTodo>? Function(Query<ReadSampleTodo> query)? queryBuilder,
    int Function(ReadSampleTodo lhs, ReadSampleTodo rhs)? compare,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    Query<ReadSampleTodo> query = readSampleTodoCollectionReference;
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    var streamQs =
        query.snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamQs = streamQs.where((qs) => !qs.metadata.hasPendingWrites);
    }
    return streamQs.map((qs) {
      final result = qs.docs.map((qds) => qds.data()).toList();
      if (compare != null) {
        result.sort(compare);
      }
      return result;
    });
  }

  /// Fetches a specific [ReadSampleTodo] document.
  Future<ReadSampleTodo?> fetchDocument({
    required String sampleTodoId,
    GetOptions? options,
  }) async {
    final ds = await readSampleTodoDocumentReference(
      sampleTodoId: sampleTodoId,
    ).get(options);
    return ds.data();
  }

  /// Subscribes a specific [SampleTodo] document.
  Stream<ReadSampleTodo?> subscribeDocument({
    required String sampleTodoId,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    var streamDs = readSampleTodoDocumentReference(
      sampleTodoId: sampleTodoId,
    ).snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamDs = streamDs.where((ds) => !ds.metadata.hasPendingWrites);
    }
    return streamDs.map((ds) => ds.data());
  }

  /// Adds a [SampleTodo] document.
  Future<DocumentReference<CreateSampleTodo>> add({
    required CreateSampleTodo createSampleTodo,
  }) =>
      createSampleTodoCollectionReference.add(createSampleTodo);

  /// Sets a [SampleTodo] document.
  Future<void> set({
    required String sampleTodoId,
    required CreateSampleTodo createSampleTodo,
    SetOptions? options,
  }) =>
      createSampleTodoDocumentReference(
        sampleTodoId: sampleTodoId,
      ).set(createSampleTodo, options);

  /// Updates a specific [SampleTodo] document.
  Future<void> update({
    required String sampleTodoId,
    required UpdateSampleTodo updateSampleTodo,
  }) =>
      updateSampleTodoDocumentReference(
        sampleTodoId: sampleTodoId,
      ).update(updateSampleTodo.toJson());

  /// Deletes a specific [SampleTodo] document.
  Future<void> delete({
    required String sampleTodoId,
  }) =>
      deleteSampleTodoDocumentReference(
        sampleTodoId: sampleTodoId,
      ).delete();
}

File: functions/src/firestore-documents/sampleTodo.flutterfireGen.ts
import * as admin from 'firebase-admin'
import {
    DocumentData,
    DocumentReference,
    DocumentSnapshot,
    FieldValue,
    Query,
    QueryDocumentSnapshot,
    QuerySnapshot,
    SetOptions,
    Timestamp,
    WriteResult
} from 'firebase-admin/firestore'

export class ReadSampleTodo {
    constructor({
        sampleTodoId,
        path,
        title,
        description,
        isDone,
        dueDateTime,
        updatedAt
    }: {
        sampleTodoId: string
        path: string
        title: string
        description: string
        isDone: boolean
        dueDateTime: Date
        updatedAt?: Date
    }) {
        this.sampleTodoId = sampleTodoId
        this.path = path
        this.title = title
        this.description = description
        this.isDone = isDone
        this.dueDateTime = dueDateTime
        this.updatedAt = updatedAt
    }

    readonly sampleTodoId: string

    readonly path: string

    readonly title: string

    readonly description: string

    readonly isDone: boolean

    readonly dueDateTime: Date

    readonly updatedAt?: Date

    private static fromJson(json: Record<string, unknown>): ReadSampleTodo {
        return new ReadSampleTodo({
            sampleTodoId: json[`sampleTodoId`] as string,
            path: json[`path`] as string,
            title: (json[`title`] as string | undefined) ?? ``,
            description: (json[`description`] as string | undefined) ?? ``,
            isDone: (json[`isDone`] as boolean | undefined) ?? false,
            dueDateTime: (json[`dueDateTime`] as Timestamp).toDate(),
            updatedAt: (json[`updatedAt`] as Timestamp | undefined)?.toDate()
        })
    }

    static fromDocumentSnapshot(ds: DocumentSnapshot): ReadSampleTodo {
        const data = ds.data()!
        const cleanedData: Record<string, unknown> = {}
        for (const [key, value] of Object.entries(data)) {
            cleanedData[key] = value === null ? undefined : value
        }
        return ReadSampleTodo.fromJson({
            ...cleanedData,
            sampleTodoId: ds.id,
            path: ds.ref.path
        })
    }
}

export class CreateSampleTodo {
    constructor({
        title,
        description,
        isDone,
        dueDateTime
    }: {
        title: string
        description: string
        isDone: boolean
        dueDateTime: Date
    }) {
        this.title = title
        this.description = description
        this.isDone = isDone
        this.dueDateTime = dueDateTime
    }

    readonly title: string

    readonly description: string

    readonly isDone: boolean

    readonly dueDateTime: Date

    toJson(): Record<string, unknown> {
        return {
            title: this.title,
            description: this.description,
            isDone: this.isDone,
            dueDateTime: this.dueDateTime,
            updatedAt: FieldValue.serverTimestamp()
        }
    }
}

export class UpdateSampleTodo {
    constructor({
        title,
        description,
        isDone,
        dueDateTime
    }: {
        title?: string
        description?: string
        isDone?: boolean
        dueDateTime?: Date
    }) {
        this.title = title
        this.description = description
        this.isDone = isDone
        this.dueDateTime = dueDateTime
    }

    readonly title?: string

    readonly description?: string

    readonly isDone?: boolean

    readonly dueDateTime?: Date

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    toJson(): Record<string, any> {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const json: Record<string, any> = {}
        if (this.title != undefined) {
            json[`title`] = this.title
        }
        if (this.description != undefined) {
            json[`description`] = this.description
        }
        if (this.isDone != undefined) {
            json[`isDone`] = this.isDone
        }
        if (this.dueDateTime != undefined) {
            json[`dueDateTime`] = this.dueDateTime
        }
        json[`updatedAt`] = FieldValue.serverTimestamp()
        return json
    }
}

export class DeleteSampleTodo {}

/**
 * A Cloud Firestore object which ignores `undefined` properties.
 */
const db = admin.firestore()
db.settings({ ignoreUndefinedProperties: true })

/**
 * Provides a reference to the sampleTodos collection for reading.
 */
export const readSampleTodoCollectionReference = db
    .collection(`sampleTodos`)
    .withConverter<ReadSampleTodo>({
        fromFirestore: (ds: DocumentSnapshot): ReadSampleTodo => {
            return ReadSampleTodo.fromDocumentSnapshot(ds)
        },
        toFirestore: () => {
            throw new Error(`toFirestore is not implemented for ReadSampleTodo`)
        }
    })

/**
 * Provides a reference to a sampleTodo document for reading.
 * @param sampleTodoId - The ID of the sampleTodo document to read.
 */
export const readSampleTodoDocumentReference = ({
    sampleTodoId
}: {
    sampleTodoId: string
}): DocumentReference<ReadSampleTodo> =>
    readSampleTodoCollectionReference.doc(sampleTodoId)

/**
 * Provides a reference to the sampleTodos collection for creating.
 */
export const createSampleTodoCollectionReference = db
    .collection(`sampleTodos`)
    .withConverter<CreateSampleTodo>({
        fromFirestore: () => {
            throw new Error(
                `fromFirestore is not implemented for CreateSampleTodo`
            )
        },
        toFirestore: (obj: CreateSampleTodo): DocumentData => {
            return obj.toJson()
        }
    })

/**
 * Provides a reference to a sampleTodo document for creating.
 * @param sampleTodoId - The ID of the sampleTodo document to read.
 */
export const createSampleTodoDocumentReference = ({
    sampleTodoId
}: {
    sampleTodoId: string
}): DocumentReference<CreateSampleTodo> =>
    createSampleTodoCollectionReference.doc(sampleTodoId)

/**
 * Provides a reference to the sampleTodos collection for updating.
 */
export const updateSampleTodoCollectionReference = db
    .collection(`sampleTodos`)
    .withConverter<UpdateSampleTodo>({
        fromFirestore: () => {
            throw new Error(
                `fromFirestore is not implemented for UpdateSampleTodo`
            )
        },
        toFirestore: (obj: UpdateSampleTodo): DocumentData => {
            return obj.toJson()
        }
    })

/**
 * Provides a reference to a sampleTodo document for updating.
 * @param sampleTodoId - The ID of the sampleTodo document to read.
 */
export const updateSampleTodoDocumentReference = ({
    sampleTodoId
}: {
    sampleTodoId: string
}): DocumentReference<UpdateSampleTodo> =>
    updateSampleTodoCollectionReference.doc(sampleTodoId)

/**
 * Provides a reference to the sampleTodos collection for deleting.
 */
export const deleteSampleTodoCollectionReference = db
    .collection(`sampleTodos`)
    .withConverter<DeleteSampleTodo>({
        fromFirestore: () => {
            throw new Error(
                `fromFirestore is not implemented for DeleteSampleTodo`
            )
        },
        toFirestore: (): DocumentData => {
            throw new Error(
                `toFirestore is not implemented for DeleteSampleTodo`
            )
        }
    })

/**
 * Provides a reference to a sampleTodo document for deleting.
 * @param sampleTodoId - The ID of the sampleTodo document to read.
 */
export const deleteSampleTodoDocumentReference = ({
    sampleTodoId
}: {
    sampleTodoId: string
}): DocumentReference<DeleteSampleTodo> =>
    deleteSampleTodoCollectionReference.doc(sampleTodoId)

/**
 * Manages queries against the sampleTodos collection.
 */
export class SampleTodoQuery {
    /**
     * Fetches sampleTodo documents.
     * @param queryBuilder - Function to modify the query.
     * @param compare - Function to sort the results.
     */
    async fetchDocuments({
        queryBuilder,
        compare
    }: {
        queryBuilder?: (query: Query<ReadSampleTodo>) => Query<ReadSampleTodo>
        compare?: (lhs: ReadSampleTodo, rhs: ReadSampleTodo) => number
    }): Promise<ReadSampleTodo[]> {
        let query: Query<ReadSampleTodo> = readSampleTodoCollectionReference
        if (queryBuilder != undefined) {
            query = queryBuilder(query)
        }
        const qs: QuerySnapshot<ReadSampleTodo> = await query.get()
        let result = qs.docs.map((qds: QueryDocumentSnapshot<ReadSampleTodo>) =>
            qds.data()
        )
        if (compare != undefined) {
            result = result.sort(compare)
        }
        return result
    }

    /**
     * Fetches a specific sampleTodo document.
     * @param sampleTodoId - The ID of the sampleTodo document to fetch.
     */
    async fetchDocument({
        sampleTodoId
    }: {
        sampleTodoId: string
    }): Promise<ReadSampleTodo | undefined> {
        const ds = await readSampleTodoDocumentReference({
            sampleTodoId
        }).get()
        return ds.data()
    }

    /**
     * Adds a sampleTodo document.
     * @param createSampleTodo - The sampleTodo details to add.
     */
    async add({
        createSampleTodo
    }: {
        createSampleTodo: CreateSampleTodo
    }): Promise<DocumentReference<CreateSampleTodo>> {
        return createSampleTodoCollectionReference.add(createSampleTodo)
    }

    /**
     * Sets a sampleTodo document.
     * @param sampleTodoId - The ID of the sampleTodo document to set.
     * @param createSampleTodo - The sampleTodo details to set.
     * @param options - Options for the set operation.
     */
    async set({
        sampleTodoId,
        createSampleTodo,
        options
    }: {
        sampleTodoId: string
        createSampleTodo: CreateSampleTodo
        options?: SetOptions
    }): Promise<WriteResult> {
        if (options == undefined) {
            return createSampleTodoDocumentReference({
                sampleTodoId
            }).set(createSampleTodo)
        } else {
            return createSampleTodoDocumentReference({
                sampleTodoId
            }).set(createSampleTodo, options)
        }
    }

    /**
     * Updates a specific sampleTodo document.
     * @param sampleTodoId - The ID of the sampleTodo document to update.
     * @param updateSampleTodo - The details for updating the sampleTodo.
     */
    async update({
        sampleTodoId,
        updateSampleTodo
    }: {
        sampleTodoId: string
        updateSampleTodo: UpdateSampleTodo
    }): Promise<WriteResult> {
        return updateSampleTodoDocumentReference({
            sampleTodoId
        }).update(updateSampleTodo.toJson())
    }

    /**
     * Deletes a specific sampleTodo document.
     * @param sampleTodoId - The ID of the sampleTodo document to delete.
     */
    async delete({
        sampleTodoId
    }: {
        sampleTodoId: string
    }): Promise<WriteResult> {
        return deleteSampleTodoDocumentReference({
            sampleTodoId
        }).delete()
    }
}
